tests/components/roomba/test_config_flow.py

```diff
@@ -0,0 +1,90 @@
+"""Test the iRobot Roomba config flow."""
+from asynctest import patch
+
+from homeassistant import config_entries, setup
+from homeassistant.components.roomba.config_flow import CannotConnect, InvalidAuth
+from homeassistant.components.roomba.const import DOMAIN
+
+
+async def test_form(hass):
+    """Test we get the form."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.roomba.config_flow.PlaceholderHub.authenticate",
```

Comment:
This needs to patch `roombapy`

Reply:
@bdraco I'm not familiar with the topic, just tell what should be added in roombapy lib?

Reply:
It looks like you'll need typed exceptions in `roomapy` 
One for invalid auth details, and one for connect connect

It looks like
https://github.com/pschmitt/roombapy/blob/master/roomba/roomba.py#L243
Is checking for rc == 0

Can you throw a `RoombaInvalidAuth`  or similar exception here when you get rc=4?

https://pypi.org/project/paho-mqtt/#client
```
0: Connection successful 1: Connection refused - incorrect protocol version 2: Connection refused - invalid client identifier 3: Connection refused - server unavailable 4: Connection refused - bad username or password 5: Connection refused - not authorised 6-255: Currently unused.
```

Reply:
yeah, sure  pschmitt/roombapy#26

Reply:
checked out throwing exception in on_connect method.
unfortunately this can't be done, because [here](https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py#L3048)
paho-mqtt just logging exception and not delegate it higher.
so, if I add throwing exception in on_connect methods it will be invisible for everyone

Reply:
Is that upstream responsive to a PR to improve the handling there?

Reply:
I created branch with exceptions, but I don't think it will be merged.
@pschmitt maybe you have some ideas?